coriolis/tests/osmorphing/osdetect/test_windows.py

```diff
@@ -0,0 +1,192 @@
+from coriolis.osmorphing.osdetect import windows
+from coriolis.tests import test_base
+
+
+@ddt.ddt
+class WindowsOSDetectToolsTestCase(test_base.CoriolisBaseTestCase):
+    """Test suite for the WindowsOSDetectTools class."""
+
+    def setUp(self):
+        super(WindowsOSDetectToolsTestCase, self).setUp()
+        self.conn = mock.MagicMock()
+        self.os_root_dir = 'C:\\'
+
+        self.windows_os_detect_tools = windows.WindowsOSDetectTools(
+            self.conn, self.os_root_dir, mock.sentinel.operation_timeout)
+
+    def test_returned_detected_os_info_fields(self):
+        expected_base_fields = [
+            "os_type",
+            "distribution_name",
+            "release_version",
+            "friendly_release_name",
+            "version_number",
+            "edition_id",
+            "installation_type",
+            "product_name"
+        ]
+
+        result = (
+            windows.WindowsOSDetectTools.returned_detected_os_info_fields()
+        )
+
+        self.assertEqual(result, expected_base_fields)
+
+    def test__load_registry_hive(self):
+        self.windows_os_detect_tools._load_registry_hive(
+            mock.sentinel.subkey, mock.sentinel.path)
+
+        self.conn.exec_command.assert_called_once_with(
+            "reg.exe", ["load", mock.sentinel.subkey, mock.sentinel.path])
+
+    def test__unload_registry_hive(self):
+        self.windows_os_detect_tools._unload_registry_hive(
+            mock.sentinel.subkey)
+
+        self.conn.exec_command.assert_called_once_with(
+            "reg.exe", ["unload", mock.sentinel.subkey])
+
+    @mock.patch.object(windows.re, 'search')
+    def test__get_ps_fl_value(self, mock_search):
+        result = self.windows_os_detect_tools._get_ps_fl_value(
+            mock.sentinel.data, mock.sentinel.name)
+
+        mock_search.assert_called_once_with(
+            r'^%s\s*: (.*)$' % mock.sentinel.name,
+            mock.sentinel.data, windows.re.MULTILINE)
+
+        self.assertEqual(result, mock_search.return_value.groups()[0])
+
+    @ddt.data(
+        {
+            'CurrentVersion': mock.sentinel.CurrentVersion,
+            'CurrentMajorVersionNumber': (
+                mock.sentinel.CurrentMajorVersionNumber),
+            'should_raise': False
+        },
+        {
+            'CurrentVersion': None,
+            'CurrentMajorVersionNumber': None,
+            'should_raise': True
+        },
+        {
+            'CurrentVersion': mock.sentinel.CurrentVersion,
+            'CurrentMajorVersionNumber': None,
+            'should_raise': False
+        }
+    )
+    @mock.patch.object(windows.WindowsOSDetectTools, '_load_registry_hive')
+    @mock.patch.object(windows.WindowsOSDetectTools, '_unload_registry_hive')
+    @mock.patch.object(windows.WindowsOSDetectTools, '_get_ps_fl_value')
+    @mock.patch.object(windows.version, 'LooseVersion')
+    @mock.patch.object(windows.uuid, 'uuid4')
+    def test__get_image_version_info(
```

Comment:
I believe this method is mocking unnecessary stuff.

The test case should work with mocking only:
* `uuid.uuid4` (so you can do proper assertions) 
* `_load_registry_hive`
* `_unload_registry_hive`
* `self._conn.exec_ps_command`

And you make `self_conn.exec_ps_command(...)` return an actual value of the `Get-ItemProperty ...` PowerShell command.

For example, this would be a proper output of that PowerShell command on Windows Server 2022:
```powershell
CurrentVersion            : 6.3
CurrentMajorVersionNumber : 10
CurrentMinorVersionNumber : 0
CurrentBuildNumber        : 20348
InstallationType          : Server
ProductName               : Windows Server 2022 Datacenter Evaluation
EditionID                 : ServerDatacenterEval
```

I suggest adding the expected PowerShell command output at the beginning of the file as multi-line string:
```python
WIN_VERSION_PS_OUTPUT = """
CurrentVersion            : 6.3
CurrentMajorVersionNumber : 10
CurrentMinorVersionNumber : 0
CurrentBuildNumber        : 20348
InstallationType          : Server
ProductName               : Windows Server 2022 Datacenter Evaluation
EditionID                 : ServerDatacenterEval
"""
```

This would be used for the test case when code-path does everything as expected.
Also, I suggest adding more dummy outputs (similar with this one), so all codepaths are tested.

This output can be easily used for the `test__get_ps_fl_value` test case. For example:

```python
result = self.windows_os_detect_tools._get_ps_fl_value(WIN_VERSION_PS_OUTPUT, "CurrentBuildNumber")
self.assertEqual(result, "20348")
```

Reply:
So nice, the test case looks much better as you suggested.